src/main/java/org/devcon/ticket/UnpredictableNumberTool.java

```diff
@@ -0,0 +1,81 @@
+package org.devcon.ticket;
+
+import com.alphawallet.attestation.core.ExceptionUtil;
+import com.alphawallet.attestation.core.URLUtility;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.time.Clock;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.bouncycastle.crypto.digests.KeccakDigest;
+import org.bouncycastle.crypto.macs.HMac;
+import org.bouncycastle.crypto.params.KeyParameter;
+import org.tokenscript.eip712.Eip712Common;
+
+public class UnpredictableNumberTool {
+  private static final Logger logger = LogManager.getLogger(UnpredictableNumberTool.class);
+  public static final long DEFAULT_VALIDITY_IN_MS = 3600*1000;
+  public static final int BYTES_IN_UN = 8; // 64 bits
+
+  private static final ByteBuffer longBuffer = ByteBuffer.allocate(Long.BYTES);
+  private final String domain;
+  private final long validityInMs;
+  private final HMac hmac = new HMac(new KeccakDigest(256));
+
+  public UnpredictableNumberTool(byte[] key, String domain) {
+    this(key, domain, DEFAULT_VALIDITY_IN_MS);
+  }
+
+  public String getDomain() {
+    return domain;
+  }
+
+  public UnpredictableNumberTool(byte[] key, String domain, long validityInMs) {
+    this.domain = domain.toLowerCase();
+    this.validityInMs = validityInMs;
+    hmac.init(new KeyParameter(key));
+    // todo should be moved to url utility
+    if (!Eip712Common.isDomainValid(domain)) {
+      ExceptionUtil.throwException(logger, new IllegalArgumentException("Domain is not a valid domain"));
+    }
+  }
+
+  /**
+   * Computes an URL friendly base 64 encoding of HMAC(expirationInMs || domain).
+   * The encoding consists of BYTES_IN_UN bytes
+   * ExpirationInMs is computed from current time plus DEFAULT_VALIDITY_IN_MS
+   */
+  public UnpredictableNumberBundle getUnpredictableNumberBundle() {
+    long expiration = Clock.systemUTC().millis() + validityInMs;
+    return new UnpredictableNumberBundle(getUnpredictableNumber(expiration), domain, expiration);
+  }
+
+  private String getUnpredictableNumber(long expirationInMs) {
+    hmac.reset();
+    hmac.update(longToBytes(expirationInMs), 0, Long.BYTES);
+    hmac.update(domain.getBytes(StandardCharsets.UTF_8), 0, domain.getBytes(StandardCharsets.UTF_8).length);
+    byte[] digest = new byte[256 / 8];
```

Comment:
Can you  generate random number , add it to hmac and prepend that UN with that number?

Like that logic:
expiry: now + 3600
UN1: rand()
UN2: hmac(concat(UN1, expiry), bogotá-secret)
UN: concat(UN1, UN2)

Reply:
I have added randomness (and kept the domain name).